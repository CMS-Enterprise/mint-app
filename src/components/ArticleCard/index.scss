.article {

    &__body p {
        display: -webkit-box;
        -webkit-line-clamp: 5;
        -webkit-box-orient: vertical;
        overflow: hidden;
        margin-bottom: .75rem;
    }
    &-card--isLink {
      cursor: pointer;
    }
    &__footer {
      bottom: 0;
    }
}
