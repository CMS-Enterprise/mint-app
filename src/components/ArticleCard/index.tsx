import React from 'react';
import { useTranslation } from 'react-i18next';
import { useHistory } from 'react-router-dom';
import {
  Card,
  CardBody,
  CardFooter,
  CardHeader
} from '@trussworks/react-uswds';
import classnames from 'classnames';

import HelpAndKnowledgeCategoryTag from 'components/HelpAndKnowledgeCategoryTag';
import UswdsReactLink from 'components/LinkWrapper';
import { ArticleTypeProps } from 'views/HelpAndKnowledge/Articles';

import './index.scss';

type ArticleCardProps = {
  className?: string;
  route: string;
  translation: string;
  isLink?: boolean;
  tag?: boolean;
};

const ArticleCard = ({
  className,
  type,
  route,
  translation,
  isLink = false,
  tag = true
}: ArticleCardProps & ArticleTypeProps) => {
  const { t } = useTranslation(translation);
  const history = useHistory();

  const clickHandler = (e: React.MouseEvent<HTMLElement>, url: string) => {
    const target = e.target as Element;
    if (isLink && target.getAttribute('data-testid') !== 'tag') {
      history.push(`/help-and-knowledge${url}`);
    }
  };

  return (
    <Card
      containerProps={{
        className: 'radius-md shadow-2 minh-mobile padding-3'
      }}
      data-testid="article-card"
      className={classnames('desktop:grid-col-4', 'article', className, {
        'article-card--isLink': isLink
      })}
      onClick={e => clickHandler(e, route)}
    >
      <CardHeader className="padding-0">
        <h3 className="line-height-body-4 margin-bottom-1">{t('title')}</h3>
      </CardHeader>
      {tag && <HelpAndKnowledgeCategoryTag type={type} />}

      <CardBody className="padding-x-0 article__body">
        <p>{t('description')}</p>
      </CardBody>
      <CardFooter className="padding-x-0 padding-top-2 padding-bottom-0">
        <UswdsReactLink
          to="#"
          className="usa-button usa-button--outline"
          variant="unstyled"
        >
          {useTranslation('helpAndKnowledge').t('read')}
        </UswdsReactLink>
      </CardFooter>
    </Card>
  );
};

export default ArticleCard;
