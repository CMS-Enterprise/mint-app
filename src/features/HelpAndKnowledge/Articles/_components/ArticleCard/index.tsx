import React from 'react';
import { useTranslation } from 'react-i18next';
import { useNavigate } from 'react-router-dom';
import {
  Card,
  CardBody,
  CardFooter,
  CardHeader
} from '@trussworks/react-uswds';
import classnames from 'classnames';
import { ArticleCategories } from 'features/HelpAndKnowledge/Articles';
import HelpCategoryTag from 'features/HelpAndKnowledge/Articles/_components/HelpCategoryTag';

import './index.scss';

type ArticleCardProps = {
  className?: string;
  route: string;
  translation: string;
  isLink?: boolean;
  tag?: boolean;
  type: ArticleCategories;
};

const ArticleCard = ({
  className,
  type,
  route,
  translation,
  isLink = false,
  tag = true
}: ArticleCardProps) => {
  const { t } = useTranslation(translation);
  const navigate = useNavigate();

  const clickHandler = (e: React.MouseEvent<HTMLElement>, url: string) => {
    const target = e.target as Element;
    if (isLink && target.getAttribute('data-testid') !== 'tag') {
      navigate(`/help-and-knowledge${url}`);
    }
  };

  return (
    <Card
      containerProps={{
        className: 'radius-md shadow-2 minh-mobile padding-3'
      }}
      data-testid="article-card"
      className={classnames('desktop:grid-col-4', 'article', className, {
        'article-card--isLink': isLink
      })}
      onClick={(e: React.MouseEvent<HTMLElement>) => clickHandler(e, route)}
    >
      <CardHeader className="padding-0">
        <h3 className="line-height-body-4 margin-bottom-1">{t('title')}</h3>
      </CardHeader>
      {tag && <HelpCategoryTag type={type} />}

      <CardBody className="padding-x-0 article__body padding-top-2">
        <p>{t('description')}</p>
      </CardBody>

      <CardFooter className="padding-x-0 padding-top-2 padding-bottom-0">
        <button
          type="button"
          className="usa-button usa-button--outline"
          aria-label={`${useTranslation('helpAndKnowledge').t('read')} ${t(
            'title'
          )} article`}
        >
          {useTranslation('helpAndKnowledge').t('read')}
        </button>
      </CardFooter>
    </Card>
  );
};

export default ArticleCard;
