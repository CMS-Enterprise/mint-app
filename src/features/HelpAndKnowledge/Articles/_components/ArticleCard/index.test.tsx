import React from 'react';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import { render } from '@testing-library/react';
import helpAndKnowledgeArticles from 'features/HelpAndKnowledge/Articles';

import ArticleCard from './index';

describe('RelatedArticle', () => {
  it('matches the snapshot', () => {
    const router = createMemoryRouter(
      [
        {
          path: '/',
          element: (
            <ArticleCard {...helpAndKnowledgeArticles[0]} key="article-1" />
          )
        }
      ],
      {
        initialEntries: ['/']
      }
    );

    const { asFragment } = render(<RouterProvider router={router} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders Article Card entirely wrapped as a link', () => {
    const router = createMemoryRouter(
      [
        {
          path: '/',
          element: (
            <ArticleCard
              {...helpAndKnowledgeArticles[0]}
              isLink
              key="article-2"
            />
          )
        }
      ],
      {
        initialEntries: ['/']
      }
    );

    const { container } = render(<RouterProvider router={router} />);

    expect(
      container.getElementsByClassName('article-card--isLink').length
    ).toBe(1);
  });
});
