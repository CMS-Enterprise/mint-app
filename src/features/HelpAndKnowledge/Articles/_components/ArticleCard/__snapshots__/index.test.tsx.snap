// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RelatedArticle > matches the snapshot 1`] = `
<DocumentFragment>
  <li
    class="usa-card desktop:grid-col-4 article"
    data-testid="article-card"
  >
    <div
      class="usa-card__container radius-md shadow-2 minh-mobile padding-3"
    >
      <div
        class="usa-card__header padding-0"
        data-testid="CardHeader"
      >
        <h3
          class="line-height-body-4 margin-bottom-1"
        >
          Evaluate your data exchange approach
        </h3>
      </div>
      <a
        class="usa-link width-fit-content display-block "
        data-discover="true"
        href="/help-and-knowledge/articles?category=getting-started"
      >
        <span
          aria-label="Articles under the Getting started category"
          class="mint-tag article__tag bg-primary-lighter text-primary text-no-uppercase text-bold font-body-sm"
          data-testid="tag"
        >
          Getting started
        </span>
      </a>
      <div
        class="usa-card__body padding-x-0 article__body padding-top-2"
        data-testid="CardBody"
      >
        <p>
          Exchanging data is an essential component of model operations and can take many forms depending on policy goals, implementation needs, and participants in the model. Examples of data that are typically exchanged include CMS sharing data with participants (for example, claims data) or participants sharing quality data (for example, a new measure or patient-reported outcome). We often collect health-related social needs data, such as socio-demographic data. Some models may have more complex needs or could be avenues to explore broader HHS policy initiatives around data exchange.
        </p>
      </div>
      <div
        class="usa-card__footer padding-x-0 padding-top-2 padding-bottom-0"
        data-testid="CardFooter"
      >
        <button
          aria-label="Read Evaluate your data exchange approach article"
          class="usa-button usa-button--outline"
          type="button"
        >
          Read
        </button>
      </div>
    </div>
  </li>
</DocumentFragment>
`;
